### PR TITLE
[OptApp] Add entity node entity filtering

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This adds the most basic filtering mechanism which is entity node entity filtering method as an additional method for existing filtering mechanism.

This PR introduces `Filter` base class which is used as the parent class for all the filtering mechanisms in OptApp.

**🆕 Changelog**
- Adds `Filter` base class
- Adds `EntityNodeEntityFilter`

### CODE REVIEW

File: applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py [hunk 1]
```diff
@@ -53,6 +53,32 @@ def test_ElementExplicitFilterConsistency(self):
...
+            self.assertVectorAlmostEqual(filtered_field[i, :], Kratos.Array3([2.0, 2.0, 2.0]) / entity.GetGeometry().DomainSize())
+
+    def test_ElementNodeElementFilter(self):
+        constant_field_value = Kratos.Array3([2.0, 2.0, 2.0])
+        element_expression = Kratos.Expression.ElementExpression(self.model_part)
+        Kratos.Expression.LiteralExpressionIO.SetData(element_expression, constant_field_value)
+        ene_filter = KratosOA.ElementNodeElementFilter(self.model_part)
+
+        filtered_field = ene_filter.FilterField(element_expression)
+        self.assertAlmostEqual(KratosOA.ExpressionUtils.NormL2(filtered_field - element_expression), 1e-9, 8)
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py]: I don't understand why filtered_field should be equal to element_expression?

File: applications/OptimizationApplication/custom_utilities/filtering/filter.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    ///@{
+
+    Filter() = default;
+
+    virtual ~Filter() = default;
+
+    ///@}
+    ///@name Public operations
+
+    virtual void Update()
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/filtering/filter.h:72]: Can you add a brief function description, what they should do?

File: applications/OptimizationApplication/custom_python/add_custom_filters_to_python.cpp [hunk 3]
```diff
@@ -35,22 +37,53 @@ void AddExplicitFilter(
...
-        .def("Update", &ExplicitFilter<TContainerType>::Update)
-        .def("__str__", &ExplicitFilter<TContainerType>::Info)
         ;
 }
 
 } // namespace Detail
 
 void AddCustomFiltersToPython(pybind11::module& m)
 {
+    namespace py = pybind11;
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_python/add_custom_filters_to_python.cpp:51]: Does it in conflict with @matekelemen 's changes to new pybind version?

File: applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py [hunk 4]
```diff
@@ -11,73 +12,51 @@ class TestExplicitFilter(kratos_unittest.TestCase):
...
-        Kratos.Expression.LiteralExpressionIO.SetData(integration_weights, constant_field_value)
-        vm_filter.GetIntegrationWeights(integration_weights)
-        integrated_constant_field = integration_weights * unfiltered_field
-        filtered_integrated_data = vm_filter.FilterIntegratedField(integrated_constant_field)
-        self.assertAlmostEqual(Kratos.Expression.Utils.NormL2(filtered_integrated_data), math.sqrt(12 * len(entities)), 12)
+        cls.model_part.ProcessInfo[Kratos.DOMAIN_SIZE] = 3
+        cls.model_part.AddNodalSolutionStepVariable(Kratos.NORMAL)
+        ReadModelPart("shell", cls.model_part)
+
+    def test_Hello(self):
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py:19]: give a proper name, please

File: applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py [hunk 5]
```diff
@@ -11,73 +12,51 @@ class TestExplicitFilter(kratos_unittest.TestCase):
...
+            Kratos.Expression.NodalPositionExpressionIO.Write(final_pos, Kratos.Configuration.Current)
+
+            explicit_filter.Update()
+
+            vtu_output = Kratos.VtuOutput(self.model_part, binary_output=Kratos.VtuOutput.ASCII)
+            vtu_output.AddContainerExpression("raw", nodal_expression.Clone())
+            vtu_output.AddContainerExpression("filtered", filtered_nodal_expression.Clone())
+            vtu_output.AddContainerExpression("normalized_filtered", normalized_filtered.Clone())
+            vtu_output.AddContainerExpression("update", update.Clone())
+            vtu_output.PrintOutput(f"output_{i}")
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py:59]: I guess you are missing comparison the results with given reference?

Comment (Igarizza) [reply applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py:59]: Damping test is missing. Also, we need a unit test for all kernel functions.

File: applications/OptimizationApplication/python_scripts/filtering/filtering_factory.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import FilterTypes
+
+def Factory(model: Kratos.Model, filtering_model_part_name: str, variable: SupportedSensitivityFieldVariableTypes, data_location: Kratos.Globals.DataLocation, settings: Kratos.Parameters) -> FilterTypes:
+    if not settings.Has("filter_type"):
+        raise RuntimeError(f"\"filter_type\" not specified in settings:\n{settings}")
+
+    if not settings.Has("filtering_model_part_name"):
+        settings.AddString("filtering_model_part_name", filtering_model_part_name)
+
+    if settings["filtering_model_part_name"].GetString() != filtering_model_part_name:
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/filtering/filtering_factory.py:15]: why do you need to give filtering_model_part_name two times, in settings and as variable?

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 7]
```diff
@@ -20,54 +20,6 @@ def Factory(model: Kratos.Model, parameters: Kratos.Parameters, optimization_pro
...
-                    "variable_name"   : "HELMHOLTZ_SCALAR",
-                    "constrained"     : true
-                }
-            }
-            """)
-        shell_scalar_filter_parameters["processes"]["boundary_conditions_process_list"].Append(auto_process_settings)
-
-    return shell_scalar_filter_parameters
-
 class ShellThicknessControl(Control):
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:23]: Do we have test for it?

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 8]
```diff
@@ -84,58 +36,48 @@ def __init__(self, name: str, model: Kratos.Model, parameters: Kratos.Parameters
...
             raise RuntimeError(f"No model parts were provided for ShellThicknessControl. [ control name = \"{self.GetName()}\"]")
 
-        self.model_part_operation = ModelPartOperation(self.model, ModelPartOperation.OperationType.UNION, f"control_{self.GetName()}", controlled_model_names_parts, False)
+        self.model_part_operation = ModelPartOperation(self.model, ModelPartOperation.OperationType.UNION, f"control_{self.GetName()}", self.parameters["controlled_model_part_names"].GetStringArray(), False)
         self.model_part: 'Optional[Kratos.ModelPart]' = None
 
-        self.fixed_model_parts_and_thicknesses: 'dict[str, float]' = {}
-        for model_part_name, thickness_params in self.parameters["fixed_model_parts_and_thicknesses"].items():
-            self.fixed_model_parts_and_thicknesses[model_part_name] = thickness_params.GetDouble()
+        self.filter = FilterFactory(self.model, self.model_part_operation.GetModelPartFullName(), Kratos.THICKNESS, Kratos.Globals.DataLocation.Element, self.parameters["filter_settings"])
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:61]: nice=)

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 9]
```diff
@@ -205,53 +144,53 @@ def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
         if not IsSameContainerExpression(new_control_field, self.GetEmptyField()):
             raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {new_control_field.GetModelPart().FullName()} ]")
 
-        if Kratos.Expression.Utils.NormL2(self.control_field - new_control_field) > 1e-15:
+        update = new_control_field - self.control_field
+        if Kratos.Expression.Utils.NormL2(update) > 1e-15:
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:148]: use isclose()

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 10]
```diff
@@ -205,53 +144,53 @@ def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
...
             with TimeLogger(self.__class__.__name__, f"Updating {self.GetName()}...", f"Finished updating of {self.GetName()}.",False):
                 # update the control thickness field
                 self.control_field = new_control_field
                 # now update the physical field
-                self._UpdateAndOutputFields()
+                self._UpdateAndOutputFields(update)
+
+                if self.beta_adaptive:
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:155]: Can you please wrap this if block to function __UpdateBeta() ?? I know it is not your implementation, but I hate long spaghetti codes...

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 11]
```diff
@@ -205,53 +144,53 @@ def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
...
                 return True
         return False
 
-    def _UpdateAndOutputFields(self) -> None:
-        # apply boundary conditions for the forward filtering
-        self._SetFixedModelPartValues()
+    def _UpdateAndOutputFields(self, update: ContainerExpressionTypes) -> None:
         # filter the control field
-        filtered_thickness_field = self.filter.FilterField(self.control_field)
+        filtered_thickness_field_update = self.filter.FilterField(update)
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:167]: this looks good

File: applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp [hunk 12]
```diff
@@ -218,22 +220,22 @@ void LinearStrainEnergyResponseUtils::CalculateStrainEnergyEntitySemiAnalyticSha
             pThreadLocalEntity->CalculateRightHandSide(rPerturbedRHS, r_process_info);
             r_initial_coordintes[0] -= Delta;
             r_coordinates[0] -= Delta;
-            AtomicAdd<double>(r_orig_node_sensitivity[0], 0.5 * inner_prod(rX, rPerturbedRHS - rRefRHS) / Delta);
+            AtomicAdd<double>(r_orig_node_sensitivity[0], inner_prod(rX, rPerturbedRHS - rRefRHS) / Delta);
```
Comment (Igarizza) [inline applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp:223]: what is problem here?

**Review by Igarizza — COMMENTED** [review]
Nice!!! very cool!!!

**Review by Igarizza — COMMENTED** [review]
Nice!!! very cool!!!

File: applications/OptimizationApplication/custom_python/add_custom_filters_to_python.cpp (hunk 3 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_python/add_custom_filters_to_python.cpp:51]: no :)

File: applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py (hunk 4 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/tests/filtering/explicit_filters_tests.py:19]: sorry for not informing, this branch was a testing branch. I created the fresh branch in #12142

File: applications/OptimizationApplication/python_scripts/filtering/filtering_factory.py (hunk 6 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/filtering/filtering_factory.py:15]: this is removed. sorry for not informing, this branch was a testing branch. I created the fresh branch in #12142

File: applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp (hunk 12 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/custom_utilities/response/linear_strain_energy_response_utils.cpp:223]: removed. sorry for not informing, this branch was a testing branch. I created the fresh branch in #12142

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py (hunk 10 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:155]: done :)

**Pull request closed without merging**